R/methods-Megastudy.R

```diff
@@ -0,0 +1,123 @@
+#' @return character
+#' @export
+setGeneric("getVariableSpecAsString", 
+  function(object) standardGeneric("getVariableSpecAsString"),
+  signature = "object"
+)
+
+#' @export
+setMethod('getVariableSpecAsString', signature('ANY'), function(object) {
+  if (!'variableSpec' %in% slotNames(object)) stop("Specified object does not have a `variableSpec` slot.")
+
+  return(getColName(object@variableSpec))
+})
+
+#' @export 
+setMethod('getVariableSpecAsString', signature('StudySpecificVocabulariesByVariable'), function(object) {
+  return(getVariableSpecAsString(object[[1]]))
+})
+
+#' StuydIdColName as String
+#' 
+#' This function returns the studyIdColName from an StudySpecificVocabulary
+#' 
+#' @param object veupathUtils::StudySpecificVocabulary
+#' @return character
+#' @export
+setGeneric("getStudyIdColName", 
+  function(object) standardGeneric("getStudyIdColName"),
+  signature = "object"
+)
+
+#' @export
+setMethod('getStudyIdColName', signature('StudySpecificVocabulary'), function(object) {
+  return(object@studyIdColName)
+})
+
+#' @export
+setMethod('getStudyIdColName', signature('StudySpecificVocabulariesByVariable'), function(object) {
+  #since we validate theyre all the same, can just take the first
+  return(object[[1]]@studyIdColName)
+})
+
+#' as.data.table
+#' 
+#' This function returns a data.table representation of 
+#' StudySpecificVocabulary or StudySpecificVocabularyByVariable
+#' 
+#' @return data.table
+#' @export 
+as.data.table <- makeGeneric('as.data.table', data.table::as.data.table)
+
+#' @export
+setMethod('as.data.table', signature('StudySpecificVocabulary'), function(x) {
+  .dt <- data.table::data.table('study'=x@study, 'variable'=x@vocabulary)
+  names(.dt) <- c(x@studyIdColName, getColName(x@variableSpec))
+
+  return(.dt)
+})
+
+#' @export 
+setMethod('as.data.table', signature('StudySpecificVocabulariesByVariable'), function(x) {
+  return(purrr::reduce(lapply(as.list(x), veupathUtils::as.data.table), rbind))
+})
+
+#' Impute Zeroes (on tall data)
+#' 
+#' This function returns a data.table which has explicit zero values
+#' for all expected categories of some variable of interest in a megastudy.
+#' 
+#' @param object veupathUtils::Megastudy
+#' @return data.table
+#' @include class-VariableMetadata.R
+#' @export
+setGeneric("imputeZeroes", 
+  function(object, variables) standardGeneric("imputeZeroes"),
+  signature = c("object", "variables")
+)
+
+#' @export
+setMethod('imputeZeroes', signature = c('Megastudy', 'VariableMetadata'), function (object, variables) {
+ 
+  weightingVariablesMetadata <- findWeightingVariablesMetadata(variables)
+  if (is.null(weightingVariablesMetadata)) return(object@data)
+
+  .dt <- object@data
+  vocabs <- object@studySpecificVocabularies
+  weightingVariablesMetadata <- findWeightingVariablesMetadata(variables)
+  if (length(weightingVariablesMetadata) > 1) stop("Megastudy class does not yet support imputing zeroes when there is more than one weighting variable present.")
```

Comment:
is there ever more than one weighting var?

Reply:
not yet, but in theory there could be one per entity